lib/Sema/TypeCheckDeclOverride.cpp

```diff
@@ -928,6 +931,98 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
   }
 }
 
+static GenericSignature *getNewGenericSignature(ValueDecl *base,
+                                                ValueDecl *derived) {
+  auto baseGenericCtx = base->getAsGenericContext();
+  auto &ctx = base->getASTContext();
+
+  if (!baseGenericCtx) {
+    return nullptr;
+  }
+
+  auto baseClass = base->getDeclContext()->getSelfClassDecl();
+
+  if (!baseClass) {
+    return nullptr;
+  }
+
+  auto derivedClass = derived->getDeclContext()->getSelfClassDecl();
+  auto *baseClassSig = baseClass->getGenericSignature();
+
+  if (!derivedClass) {
+    return nullptr;
+  }
+
+  if (derivedClass->getSuperclass().isNull()) {
+    return nullptr;
+  }
+
+  auto subMap = derivedClass->getSuperclass()->getContextSubstitutionMap(
+      derivedClass->getModuleContext(), baseClass);
+
+  if (baseGenericCtx->getGenericSignature() != nullptr) {
+    unsigned depth = 0;
+
+    if (auto *genericSig = baseClass->getGenericSignature())
+      depth = genericSig->getGenericParams().back()->getDepth() + 1;
+
+    GenericSignatureBuilder builder(ctx);
+    builder.addGenericSignature(derivedClass->getGenericSignature());
+
+    if (auto derivedGenericCtx = derived->getAsGenericContext()) {
+      if (derivedGenericCtx->isGeneric()) {
+        for (auto param : *derivedGenericCtx->getGenericParams()) {
```

Comment:
I think there's a subtle problem here. This should add the generic parameters of the base method, but translate them as with substFn.

Otherwise, you might get a crash or bad behavior if the override has a different number of generic parameters than the base method, since below you're adding requirements from the base signature.